source/patterns/@aws-solutions-constructs/resources/lib/key-policy-updater-custom-resource/index.ts

```diff
@@ -0,0 +1,140 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
+ *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
+ *  and limitations under the License.
+ */
+
+import { KMSClient, GetKeyPolicyCommand, DescribeKeyCommand, PutKeyPolicyCommand, KeyManagerType } from "@aws-sdk/client-kms";
+
+const kmsClient = new KMSClient();
+
+export const handler = async (event: any, context: any) => {
+
+  let status = 'SUCCESS';
+  let responseData = {};
+
+  if (event.RequestType === 'Create' || event.RequestType === 'Update') {
+
+    try {
+      const kmsKeyId = event.ResourceProperties.KmsKeyId;
+      const cloudFrontDistributionId = event.ResourceProperties.CloudFrontDistributionId;
+      const accountId = event.ResourceProperties.AccountId;
+      const region = process.env.AWS_REGION;
+
+      const describeKeyCommandResponse = await kmsClient.send(new DescribeKeyCommand({
+        KeyId: kmsKeyId
+      }));
+
+      if (describeKeyCommandResponse.KeyMetadata?.KeyManager === KeyManagerType.AWS) {
+        return {
+          Status: 'SUCCESS',
+          Reason: 'An AWS managed key was provided, no action needed from the custom resource, exiting now.',
+          PhysicalResourceId: event.PhysicalResourceId ?? context.logStreamName,
+          StackId: event.StackId,
+          RequestId: event.RequestId,
+          LogicalResourceId: event.LogicalResourceId,
+          Data: 'An AWS managed key was provided, no action needed from the custom resource, exiting now.',
+        };
+      }
+
+      // The PolicyName is specified as "default" below because that is the only valid name as
+      // written in the documentation for @aws-sdk/client-kms.GetKeyPolicyCommandInput:
+      // https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-client-kms/Interface/GetKeyPolicyCommandInput/
+      const getKeyPolicyCommandResponse = await kmsClient.send(new GetKeyPolicyCommand({
+        KeyId: kmsKeyId,
+        PolicyName: 'default'
+      }));
+
+      if (!getKeyPolicyCommandResponse.Policy) {
+        return {
+          Status: 'FAILED',
+          Reason: 'An error occurred while retrieving the key policy.',
+          PhysicalResourceId: event.PhysicalResourceId ?? context.logStreamName,
+          StackId: event.StackId,
+          RequestId: event.RequestId,
+          LogicalResourceId: event.LogicalResourceId,
+          Data: 'An error occurred while retrieving the key policy.',
+        };
+      }
+
+      // Update the existing key policy to allow the CloudFront distribution to use the key
+      const keyPolicy = JSON.parse(getKeyPolicyCommandResponse?.Policy);
+      const keyPolicyStatementSid: string = 'Grant-CloudFront-Distribution-Key-Usage';
+
+      if (checkForExistingKeyPolicyStatement(keyPolicy, keyPolicyStatementSid)) {
```

Comment:
What if the `sid` already exists, but a new CMK is used, for example? We probably wouldn't want to return here in that case. Maybe the better logic would be to always overwrite the existing statement we are looking for, even if it ends up being a no-op .

Reply:
Addressed this in https://github.com/awslabs/aws-solutions-constructs/commit/7f2454544438c90c5709e104230ae494d15e7879

Let me know if I'm missing anything with the implementation